src/main/java/fi/aalto/cs/apluscourses/intellij/actions/SubmitExerciseAction.java

```diff
@@ -41,146 +49,135 @@
-                                                        @NotNull APlusAuthentication authentication,
-                                                        @Nullable Project project) {
-    try {
-      SubmittableExercise submittableExercise = SubmittableExercise.fromExerciseId(
-          exercise.getId(), authentication);
-      if (submittableExercise.getFilenames().isEmpty()) {
-        notifier.notify(new NotSubmittableNotification(), project);
-        return null;
-      }
-      return submittableExercise;
-    } catch (IOException e) {
-      notifyNetworkError(e, project);
-      return null;
-    }
-  }
+    SubmissionHistory history = exerciseDataSource.getSubmissionHistory(exercise);
 
-  @Nullable
-  private SubmissionHistory tryGetSubmissionHistory(@NotNull Exercise exercise,
-                                                    @NotNull APlusAuthentication authentication,
-                                                    @Nullable Project project) {
-    try {
-      return SubmissionHistory.getSubmissionHistory(exercise.getId(), authentication);
-    } catch (IOException e) {
-      notifyNetworkError(e, project);
-      return null;
-    }
-  }
+    Course course = courseViewModel.getModel();
+    List<Group> groups = exerciseDataSource.getGroups(course);
 
-  @Nullable
-  private List<Group> tryGetUserGroups(@NotNull Course course,
-                                       @NotNull APlusAuthentication authentication,
-                                       @Nullable Project project) {
-    try {
-      return Group.getGroups(course, authentication);
-    } catch (IOException e) {
-      notifyNetworkError(e, project);
-      return null;
-    }
-  }
+    SubmissionViewModel submission =
+        new SubmissionViewModel(exercise, submissionInfo, history, groups, files);
 
-  @CalledWithReadLock
-  @Nullable
-  private List<Path> tryGetFilePaths(@NotNull List<String> filenames,
-                                     @NotNull Module module,
-                                     @Nullable Project project) {
-    Path modulePath = Paths.get(ModuleUtilCore.getModuleDirPath(module));
-    List<Path> filePaths = new ArrayList<>(filenames.size());
-    for (String filename : filenames) {
-      Path filePath = VfsUtil.findFileInDirectory(modulePath, filename);
-      if (filePath == null) {
-        notifier.notify(new MissingFileNotification(module.getName(), filename), project);
-        return null;
-      }
-      filePaths.add(filePath);
+    if (!dialogs.create(submission, project).showAndGet()) {
+      return;
     }
-    return filePaths;
+
+    exerciseDataSource.submit(submission.buildSubmission());
   }
 
   private void notifyNetworkError(@NotNull IOException exception, @Nullable Project project) {
     notifier.notify(new NetworkErrorNotification(exception), project);
   }
+
+  public interface ModuleSource {
+    Module[] getModules(Project project);
+  }
```

Comment:
Is this interface for testing purposes? If `SubmitExerciseActionTest` extends `LightIdeaTestCase`, then the standard way of getting modules should work and this shouldn't be needed

Reply:
On the other hand, we don't need `LightIdeaTestCase` if we have this. But if you dislike it, we can get rid of it, I think. Yes, it's for testing purposes.

Reply:
Hmm I guess my brain was still on my branch where `SubmitExerciseAction` had like 8 constructor params with all the factories. Now the situation isn't that bad anymore, so actually having this interface is not a problem for me

Reply:
I'm glad this is being discussed. @OlliKiljunen @nikke234 

The IJ Platform has a certain approach to testing (and here I mean the `Light/HeavyIdeaTestCase`s), so overall, I see little to none reason to introduce any additional code to ~~reinvent the wheel~~ overcome the challenges we face. 

In simple words, and this is my deeper position on this: any code that we write ourselves is a maintenance burden in the future, whereas any **3-rd party code** you reuse is someone else's problem (in terms of testing/maintenance etc.), so **I prefer** latter.  I'm aware there are additional factors in play here, like "control" and "flexibility", however, with the current resourcing, we are clearly not capable of producing the high level of code test coverage.

Reply:
I'm also glad this is being discussed.
 
My position is that reliances on 3rd-party code are also a source of maintenance burden, and a really major one in cases where the 3rd-party reliance cannot be encapsulated and only used via meaningful interface (examples of nicely encapsulated 3rd party reliances are how we use https and zip libraries).

IntelliJ's approach to testing is that no real _unit_ tests are used, that is, nothing is tested in isolation but in every test there's the whole platform behind. IMO, that strips unit testing of its greatest advantage, which is, that every _unit_ of software can be implemented, understood and verified in itself, without having to implement, understand and verify the whole software. That's why I dislike `IdeaTestCase`s and I'd restrict their usage in higher-than-unit level testing.